File: curriculum/10_distributed_cache/go-impl/cache.go
package distributedcache

import (
	"context"
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"
)

var (
	ErrInvalidKey    = errors.New("invalid key")
	ErrValueTooLarge = errors.New("value too large")
	ErrBackingStore  = errors.New("backing store unavailable")
	ErrInvalidation  = errors.New("exactly one invalidation scope is required")
)

type LoaderFunc func(context.Context, string) (string, error)
type WriterFunc func(context.Context, string, string) error

type Config struct {
	NodeID          string
	Shards          []Node
	CapacityEntries int
	MaxValueBytes   int
	EvictionPolicy  string
	VirtualNodes    int
	DefaultTTL      time.Duration
	Loader          LoaderFunc
	Writer          WriterFunc
}

type Node struct {
	ID      string `json:"nodeId"`
	Address string `json:"address"`
}

type Cache struct {
	mu      sync.RWMutex
	entries map[string]*entry
	cfg     Config
	ring    *HashRing
	metrics Metrics
	logger  *slog.Logger
	flights map[string]*flight
	closed  bool
}

type entry struct {
	Key         string
	Namespace   string
	Value       string
	CreatedAt   time.Time
	UpdatedAt   time.Time
	AccessedAt  time.Time
	ExpiresAt   time.Time
	AccessCount int
	Version     int
}

type GetResult struct {
	Key          string        `json:"key"`
	Value        string        `json:"value,omitempty"`
	Hit          bool          `json:"hit"`
	Loaded       bool          `json:"loaded"`
	Coalesced    bool          `json:"coalesced"`
	TTLRemaining time.Duration `json:"-"`
	ShardID      string        `json:"shardId,omitempty"`
	NodeID       string        `json:"nodeId,omitempty"`
	Version      int           `json:"version,omitempty"`
}

func (r GetResult) MarshalJSON() ([]byte, error) {
	type wire struct {
		Key            string `json:"key"`
		Value          string `json:"value,omitempty"`
		Hit            bool   `json:"hit"`
		Loaded         bool   `json:"loaded"`
		Coalesced      bool   `json:"coalesced"`
		TTLRemainingMs int64  `json:"ttlRemainingMs,omitempty"`
		ShardID        string `json:"shardId,omitempty"`
		NodeID         string `json:"nodeId,omitempty"`
		Version        int    `json:"version,omitempty"`
	}
	return json.Marshal(wire{r.Key, r.Value, r.Hit, r.Loaded, r.Coalesced, r.TTLRemaining.Milliseconds(), r.ShardID, r.NodeID, r.Version})
}

type SetResult struct {
	Key       string   `json:"key"`
	Stored    bool     `json:"stored"`
	ShardID   string   `json:"shardId"`
	NodeID    string   `json:"nodeId"`
	Version   int      `json:"version"`
	Evicted   []string `json:"evicted"`
	ExpiresAt string   `json:"ttlExpiresAt,omitempty"`
}

type DeleteResult struct {
	Key     string `json:"key"`
	Deleted bool   `json:"deleted"`
	Reason  string `json:"reason,omitempty"`
	ShardID string `json:"shardId,omitempty"`
	NodeID  string `json:"nodeId,omitempty"`
}

type Invalidation struct{ Key, Namespace, Prefix string }

type Metrics struct {
	Hits                  int64 `json:"hits"`
	Misses                int64 `json:"misses"`
	Evictions             int64 `json:"evictions"`
	Expirations           int64 `json:"expirations"`
	Invalidations         int64 `json:"invalidations"`
	LoaderCalls           int64 `json:"loaderCalls"`
	SingleflightCoalesces int64 `json:"singleflightCoalesces"`
	MembershipChanges     int64 `json:"membershipChanges"`
}

type flight struct {
	done    chan struct{}
	value   string
	err     error
	waiters int
}

func NewCache(cfg Config) *Cache {
	if cfg.NodeID == "" {
		cfg.NodeID = "node-a"
	}
	if len(cfg.Shards) == 0 {
		cfg.Shards = []Node{{ID: cfg.NodeID, Address: "local"}}
	}
	if cfg.CapacityEntries <= 0 {
		cfg.CapacityEntries = 128
	}
	if cfg.MaxValueBytes <= 0 {
		cfg.MaxValueBytes = 1024 * 1024
	}
	if cfg.VirtualNodes <= 0 {
		cfg.VirtualNodes = 32
	}
	if cfg.EvictionPolicy == "" {
		cfg.EvictionPolicy = "lru"
	}
	return &Cache{entries: map[string]*entry{}, cfg: cfg, ring: NewHashRing(cfg.Shards, cfg.VirtualNodes), logger: slog.Default(), flights: map[string]*flight{}}
}

func (c *Cache) Set(ctx context.Context, key, value, namespace string, ttl time.Duration, writeThrough bool) (SetResult, error) {
	if err := c.validateKeyValue(key, value); err != nil {
		return SetResult{}, err
	}
	if ttl < 0 {
		return SetResult{}, ErrInvalidKey
	}
	if writeThrough && c.cfg.Writer != nil {
		if err := c.cfg.Writer(ctx, key, value); err != nil {
			return SetResult{}, ErrBackingStore
		}
	}
	now := time.Now()
	if ttl == 0 {
		ttl = c.cfg.DefaultTTL
	}
	var expires time.Time
	if ttl > 0 {
		expires = now.Add(ttl)
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	oldVersion := 0
	if old := c.entries[key]; old != nil {
		oldVersion = old.Version
	}
	c.entries[key] = &entry{Key: key, Namespace: namespace, Value: value, CreatedAt: now, UpdatedAt: now, AccessedAt: now, ExpiresAt: expires, AccessCount: 1, Version: oldVersion + 1}
	evicted := c.evictLocked(key)
	node := c.ring.Owner(key)
	c.logger.Info("cache set", "key", key, "node", node.ID, "evicted", len(evicted))
	res := SetResult{Key: key, Stored: true, ShardID: node.ID, NodeID: node.ID, Version: oldVersion + 1, Evicted: evicted}
	if !expires.IsZero() {
		res.ExpiresAt = expires.UTC().Format(time.RFC3339)
	}
	return res, nil
}

func (c *Cache) Get(ctx context.Context, key string, loadOnMiss bool) (GetResult, error) {
	if key == "" || len(key) > 512 {
		return GetResult{}, ErrInvalidKey
	}
	if res, ok := c.getLocal(key); ok {
		return res, nil
	}
	if !loadOnMiss || c.cfg.Loader == nil {
		return GetResult{Key: key, Hit: false}, nil
	}
	return c.loadSingleflight(ctx, key)
}

func (c *Cache) getLocal(key string) (GetResult, bool) {
	now := time.Now()
	c.mu.Lock()
	defer c.mu.Unlock()
	e := c.entries[key]
	if e == nil {
		c.metrics.Misses++
		return GetResult{}, false
	}
	if !e.ExpiresAt.IsZero() && !now.Before(e.ExpiresAt) {
		delete(c.entries, key)
		c.metrics.Misses++
		c.metrics.Expirations++
		return GetResult{}, false
	}
	e.AccessedAt = now
	e.AccessCount++
	c.metrics.Hits++
	node := c.ring.Owner(key)
	remaining := time.Duration(0)
	if !e.ExpiresAt.IsZero() {
		remaining = time.Until(e.ExpiresAt)
	}
	return GetResult{Key: key, Value: e.Value, Hit: true, TTLRemaining: remaining, ShardID: node.ID, NodeID: node.ID, Version: e.Version}, true
}

func (c *Cache) loadSingleflight(ctx context.Context, key string) (GetResult, error) {
	c.mu.Lock()
	if f := c.flights[key]; f != nil {
		f.waiters++
		c.metrics.SingleflightCoalesces++
		c.mu.Unlock()
		select {
		case <-f.done:
			return c.loadedResult(key, f.value, true, f.err)
		case <-ctx.Done():
			return GetResult{}, ctx.Err()
		}
	}
	f := &flight{done: make(chan struct{})}
	c.flights[key] = f
	c.metrics.LoaderCalls++
	c.mu.Unlock()
	f.value, f.err = c.cfg.Loader(ctx, key)
	c.mu.Lock()
	delete(c.flights, key)
	c.mu.Unlock()
	if f.err == nil {
		_, f.err = c.Set(ctx, key, f.value, "", c.cfg.DefaultTTL, false)
	}
	close(f.done)
	return c.loadedResult(key, f.value, false, f.err)
}

func (c *Cache) loadedResult(key, value string, coalesced bool, err error) (GetResult, error) {
	if err != nil {
		return GetResult{}, err
	}
	node := c.ring.Owner(key)
	return GetResult{Key: key, Value: value, Hit: false, Loaded: true, Coalesced: coalesced, ShardID: node.ID, NodeID: node.ID, Version: 1, TTLRemaining: c.cfg.DefaultTTL}, nil
}

func (c *Cache) Delete(ctx context.Context, key string) (DeleteResult, error) {
	if key == "" {
		return DeleteResult{}, ErrInvalidKey
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	node := c.ring.Owner(key)
	if _, ok := c.entries[key]; ok {
		delete(c.entries, key)
		c.metrics.Invalidations++
		return DeleteResult{Key: key, Deleted: true, ShardID: node.ID, NodeID: node.ID}, nil
	}
	return DeleteResult{Key: key, Deleted: false, Reason: "not_found", ShardID: node.ID, NodeID: node.ID}, nil
}

func (c *Cache) Invalidate(scope Invalidation) (int, error) {
	selected := 0
	if scope.Key != "" {
		selected++
	}
	if scope.Namespace != "" {
		selected++
	}
	if scope.Prefix != "" {
		selected++
	}
	if selected != 1 {
		return 0, ErrInvalidation
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	deleted := 0
	for key, e := range c.entries {
		match := key == scope.Key || (scope.Namespace != "" && e.Namespace == scope.Namespace) || (scope.Prefix != "" && strings.HasPrefix(key, scope.Prefix))
		if match {
			delete(c.entries, key)
			deleted++
		}
	}
	c.metrics.Invalidations += int64(deleted)
	return deleted, nil
}

func (c *Cache) Metrics() Metrics { c.mu.RLock(); defer c.mu.RUnlock(); return c.metrics }
func (c *Cache) Ring() *HashRing  { return c.ring }
func (c *Cache) Shutdown(context.Context) error {
	c.mu.Lock()
	c.closed = true
	c.mu.Unlock()
	c.logger.Info("cache shutdown")
	return nil
}

func (c *Cache) validateKeyValue(key, value string) error {
	if key == "" || len(key) > 512 {
		return ErrInvalidKey
	}
	if len([]byte(value)) > c.cfg.MaxValueBytes {
		return ErrValueTooLarge
	}
	return nil
}

func (c *Cache) evictLocked(protected string) []string {
	evicted := []string{}
	for len(c.entries) > c.cfg.CapacityEntries {
		victim := c.victimLocked(protected)
		if victim == "" {
			break
		}
		delete(c.entries, victim)
		c.metrics.Evictions++
		evicted = append(evicted, victim)
	}
	return evicted
}

func (c *Cache) victimLocked(protected string) string {
	var victim string
	for key, e := range c.entries {
		if key == protected {
			continue
		}
		if victim == "" {
			victim = key
			continue
		}
		v := c.entries[victim]
		if c.cfg.EvictionPolicy == "lfu" {
			if e.AccessCount < v.AccessCount || (e.AccessCount == v.AccessCount && (e.AccessedAt.Before(v.AccessedAt) || (e.AccessedAt.Equal(v.AccessedAt) && key < victim))) {
				victim = key
			}
		} else if e.AccessedAt.Before(v.AccessedAt) || (e.AccessedAt.Equal(v.AccessedAt) && key < victim) {
			victim = key
		}
	}
	return victim
}

type HashRing struct {
	mu          sync.RWMutex
	replicas    int
	tokens      []uint64
	owners      map[uint64]Node
	ringVersion int
}

func NewHashRing(nodes []Node, replicas int) *HashRing {
	if replicas <= 0 {
		replicas = 32
	}
	hr := &HashRing{replicas: replicas, owners: map[uint64]Node{}, ringVersion: 1}
	for _, n := range nodes {
		hr.Add(n)
	}
	return hr
}

func (h *HashRing) Add(node Node) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if node.ID == "" {
		node.ID = node.Address
	}
	for i := 0; i < h.replicas; i++ {
		token := hash(fmt.Sprintf("%s-%d", node.ID, i))
		h.tokens = append(h.tokens, token)
		h.owners[token] = node
	}
	sort.Slice(h.tokens, func(i, j int) bool { return h.tokens[i] < h.tokens[j] })
	h.ringVersion++
}

func (h *HashRing) Owner(key string) Node {
	h.mu.RLock()
	defer h.mu.RUnlock()
	if len(h.tokens) == 0 {
		return Node{}
	}
	t := hash(key)
	i := sort.Search(len(h.tokens), func(i int) bool { return h.tokens[i] >= t })
	if i == len(h.tokens) {
		i = 0
	}
	return h.owners[h.tokens[i]]
}

func (h *HashRing) TokensFor(nodeID string) []uint64 {
	h.mu.RLock()
	defer h.mu.RUnlock()
	out := []uint64{}
	for _, t := range h.tokens {
		if h.owners[t].ID == nodeID {
			out = append(out, t)
		}
	}
	return out
}

func hash(s string) uint64 { sum := sha256.Sum256([]byte(s)); return binary.BigEndian.Uint64(sum[:8]) }

type MemoryStore struct {
	mu         sync.Mutex
	data       map[string]string
	failWrites bool
}

func NewMemoryStore(data map[string]string) *MemoryStore {
	if data == nil {
		data = map[string]string{}
	}
	return &MemoryStore{data: data}
}
func (s *MemoryStore) Load(_ context.Context, key string) (string, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	v, ok := s.data[key]
	if !ok {
		return "", ErrBackingStore
	}
	return v, nil
}
func (s *MemoryStore) Write(_ context.Context, key, value string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.failWrites {
		return ErrBackingStore
	}
	s.data[key] = value
	return nil
}
func (s *MemoryStore) FailWrites(v bool) { s.mu.Lock(); s.failWrites = v; s.mu.Unlock() }

func NewHTTPServer(c *Cache) http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, http.StatusOK, map[string]string{"status": "ok", "nodeId": c.cfg.NodeID})
	})
	mux.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) { writeJSON(w, http.StatusOK, c.Metrics()) })
	mux.HandleFunc("/cluster/ring", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, http.StatusOK, map[string]any{"ringVersion": c.ring.ringVersion, "virtualNodes": c.cfg.VirtualNodes, "nodes": c.cfg.Shards})
	})
	mux.HandleFunc("/cache/invalidate", func(w http.ResponseWriter, r *http.Request) {
		var body struct{ Key, Namespace, Prefix string }
		_ = json.NewDecoder(r.Body).Decode(&body)
		count, err := c.Invalidate(Invalidation{body.Key, body.Namespace, body.Prefix})
		if err != nil {
			writeError(w, http.StatusBadRequest, err)
			return
		}
		writeJSON(w, http.StatusAccepted, map[string]any{"accepted": true, "matchedApprox": count, "completedOnNodes": []string{c.cfg.NodeID}})
	})
	mux.HandleFunc("/cache/", func(w http.ResponseWriter, r *http.Request) {
		key := strings.TrimPrefix(r.URL.Path, "/cache/")
		switch r.Method {
		case http.MethodGet:
			res, err := c.Get(r.Context(), key, r.URL.Query().Get("loadOnMiss") == "true")
			if err != nil {
				writeError(w, http.StatusBadGateway, err)
				return
			}
			if !res.Hit && !res.Loaded {
				writeJSON(w, http.StatusNotFound, res)
				return
			}
			writeJSON(w, http.StatusOK, res)
		case http.MethodPut:
			var body struct {
				Value        string
				TTLMs        int64
				WriteThrough bool
				Namespace    string
			}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeError(w, http.StatusBadRequest, err)
				return
			}
			res, err := c.Set(r.Context(), key, body.Value, body.Namespace, time.Duration(body.TTLMs)*time.Millisecond, body.WriteThrough)
			if err != nil {
				status := http.StatusBadRequest
				if errors.Is(err, ErrValueTooLarge) {
					status = http.StatusRequestEntityTooLarge
				}
				writeError(w, status, err)
				return
			}
			status := http.StatusCreated
			if res.Version > 1 {
				status = http.StatusOK
			}
			writeJSON(w, status, res)
		case http.MethodDelete:
			res, err := c.Delete(r.Context(), key)
			if err != nil {
				writeError(w, http.StatusBadRequest, err)
				return
			}
			writeJSON(w, http.StatusOK, res)
		default:
			w.WriteHeader(http.StatusMethodNotAllowed)
		}
	})
	return mux
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}
func writeError(w http.ResponseWriter, status int, err error) {
	writeJSON(w, status, map[string]any{"code": strings.ToUpper(strings.ReplaceAll(err.Error(), " ", "_")), "message": err.Error(), "retryable": status >= 500})
}

File: curriculum/10_distributed_cache/node-impl/src/cache.ts
import { createHash } from 'node:crypto';

export enum EvictionPolicy {
  Lru = 'lru',
  Lfu = 'lfu',
}

export class CacheError extends Error {
  constructor(public readonly code: string, message: string) {
    super(message);
  }
}

export class NodeInfo {
  constructor(public readonly id: string, public readonly address = '') {}
}

export type Loader = (key: string) => Promise<string> | string;
export type Writer = (key: string, value: string) => Promise<void> | void;

export class Config {
  shards: NodeInfo[];
  capacityEntries = 128;
  maxValueBytes = 1024 * 1024;
  evictionPolicy = EvictionPolicy.Lru;
  virtualNodes = 32;
  defaultTtlMs = 0;
  loader?: Loader;
  writer?: Writer;

  constructor(public readonly nodeId: string) {
    this.shards = [new NodeInfo(nodeId)];
  }

  withCapacityEntries(value: number): this {
    this.capacityEntries = value;
    return this;
  }

  withMaxValueBytes(value: number): this {
    this.maxValueBytes = value;
    return this;
  }

  withEvictionPolicy(value: EvictionPolicy): this {
    this.evictionPolicy = value;
    return this;
  }

  withDefaultTtlMs(value: number): this {
    this.defaultTtlMs = value;
    return this;
  }

  withLoader(value: Loader): this {
    this.loader = value;
    return this;
  }

  withWriter(value: Writer): this {
    this.writer = value;
    return this;
  }
}

type Entry = {
  key: string;
  namespace?: string;
  value: string;
  createdAt: number;
  accessedAt: number;
  accessSeq: number;
  expiresAt?: number;
  accessCount: number;
  version: number;
};

export type Metrics = {
  hits: number;
  misses: number;
  evictions: number;
  expirations: number;
  invalidations: number;
  loaderCalls: number;
  singleflightCoalesces: number;
  membershipChanges: number;
};

export type GetResult = {
  key: string;
  value?: string;
  hit: boolean;
  loaded: boolean;
  coalesced: boolean;
  ttlRemainingMs: number;
  shardId: string;
  nodeId: string;
  version: number;
};

export type SetOptions = { namespace?: string; ttlMs?: number; writeThrough?: boolean };
export type SetResult = { key: string; stored: boolean; version: number; evicted: string[]; shardId: string; nodeId: string };
export type DeleteResult = { key: string; deleted: boolean; reason?: string; shardId: string; nodeId: string };

export class Invalidation {
  private constructor(readonly kind: 'key' | 'namespace' | 'prefix', readonly value: string) {}
  static key(value: string): Invalidation {
    return new Invalidation('key', value);
  }
  static namespace(value: string): Invalidation {
    return new Invalidation('namespace', value);
  }
  static prefix(value: string): Invalidation {
    return new Invalidation('prefix', value);
  }
}

type Flight = Promise<string>;

export class Cache {
  private entries = new Map<string, Entry>();
  private readonly ring: HashRing;
  private readonly inflight = new Map<string, Flight>();
  private readonly counters: Metrics = { hits: 0, misses: 0, evictions: 0, expirations: 0, invalidations: 0, loaderCalls: 0, singleflightCoalesces: 0, membershipChanges: 0 };
  private shutdownFlag = false;
  private accessSeq = 0;

  constructor(private readonly config: Config) {
    this.ring = new HashRing(config.shards, config.virtualNodes);
  }

  async set(key: string, value: string, options: SetOptions = {}): Promise<SetResult> {
    this.validate(key, value, options.ttlMs);
    if (options.writeThrough && this.config.writer) {
      await this.config.writer(key, value);
    }
    const now = Date.now();
    const ttlMs = options.ttlMs ?? this.config.defaultTtlMs;
    const old = this.entries.get(key);
    this.entries.set(key, { key, namespace: options.namespace, value, createdAt: old?.createdAt ?? now, accessedAt: now, accessSeq: this.nextAccessSeq(), expiresAt: ttlMs > 0 ? now + ttlMs : undefined, accessCount: 1, version: (old?.version ?? 0) + 1 });
    const evicted = this.evict(key);
    const owner = this.ring.owner(key);
    console.log(JSON.stringify({ level: 'info', event: 'cache_set', key, nodeId: owner.id, evicted: evicted.length }));
    return { key, stored: true, version: (old?.version ?? 0) + 1, evicted, shardId: owner.id, nodeId: owner.id };
  }

  async get(key: string, options: { loadOnMiss?: boolean } | boolean = {}): Promise<GetResult> {
    this.validateKey(key);
    const local = this.getLocal(key);
    if (local) return local;
    const loadOnMiss = typeof options === 'boolean' ? options : options.loadOnMiss === true;
    if (!loadOnMiss || !this.config.loader) return this.miss(key);
    return this.loadSingleflight(key);
  }

  async delete(key: string): Promise<DeleteResult> {
    this.validateKey(key);
    const deleted = this.entries.delete(key);
    if (deleted) this.counters.invalidations += 1;
    const owner = this.ring.owner(key);
    return { key, deleted, reason: deleted ? undefined : 'not_found', shardId: owner.id, nodeId: owner.id };
  }

  async invalidate(scope: Invalidation): Promise<number> {
    const keys = [...this.entries.values()].filter((entry) => this.matches(scope, entry)).map((entry) => entry.key);
    for (const key of keys) this.entries.delete(key);
    this.counters.invalidations += keys.length;
    return keys.length;
  }

  metrics(): Metrics {
    return { ...this.counters };
  }

  async shutdown(): Promise<void> {
    this.shutdownFlag = true;
    console.log(JSON.stringify({ level: 'info', event: 'shutdown' }));
  }

  isShutdown(): boolean {
    return this.shutdownFlag;
  }

  ringInfo(): { ringVersion: number; virtualNodes: number; nodes: NodeInfo[] } {
    return { ringVersion: this.ring.version, virtualNodes: this.config.virtualNodes, nodes: this.config.shards };
  }

  private getLocal(key: string): GetResult | undefined {
    const now = Date.now();
    const entry = this.entries.get(key);
    if (!entry) {
      this.counters.misses += 1;
      return undefined;
    }
    if (entry.expiresAt !== undefined && now >= entry.expiresAt) {
      this.entries.delete(key);
      this.counters.misses += 1;
      this.counters.expirations += 1;
      return undefined;
    }
    entry.accessedAt = now;
    entry.accessSeq = this.nextAccessSeq();
    entry.accessCount += 1;
    this.counters.hits += 1;
    const owner = this.ring.owner(key);
    return { key, value: entry.value, hit: true, loaded: false, coalesced: false, ttlRemainingMs: entry.expiresAt === undefined ? 0 : Math.max(0, entry.expiresAt - now), shardId: owner.id, nodeId: owner.id, version: entry.version };
  }

  private async loadSingleflight(key: string): Promise<GetResult> {
    const existing = this.inflight.get(key);
    if (existing) {
      this.counters.singleflightCoalesces += 1;
      const value = await existing;
      return this.loadedResult(key, value, true);
    }
    this.counters.loaderCalls += 1;
    const flight = Promise.resolve(this.config.loader?.(key)).then((value) => {
      if (value === undefined) throw new CacheError('BACKING_STORE', 'backing store unavailable');
      return value;
    });
    this.inflight.set(key, flight);
    try {
      const value = await flight;
      await this.set(key, value, { ttlMs: this.config.defaultTtlMs || undefined });
      return this.loadedResult(key, value, false);
    } finally {
      this.inflight.delete(key);
    }
  }

  private loadedResult(key: string, value: string, coalesced: boolean): GetResult {
    const owner = this.ring.owner(key);
    return { key, value, hit: false, loaded: true, coalesced, ttlRemainingMs: this.config.defaultTtlMs, shardId: owner.id, nodeId: owner.id, version: 1 };
  }

  private miss(key: string): GetResult {
    return { key, hit: false, loaded: false, coalesced: false, ttlRemainingMs: 0, shardId: '', nodeId: '', version: 0 };
  }

  private validate(key: string, value: string, ttlMs?: number): void {
    this.validateKey(key);
    if (Buffer.byteLength(value) > this.config.maxValueBytes) throw new CacheError('VALUE_TOO_LARGE', 'value too large');
    if (ttlMs !== undefined && ttlMs <= 0) throw new CacheError('INVALID_TTL', 'ttl must be positive');
  }

  private validateKey(key: string): void {
    if (key.length === 0 || Buffer.byteLength(key) > 512) throw new CacheError('INVALID_KEY', 'invalid key');
  }

  private evict(protectedKey: string): string[] {
    const evicted: string[] = [];
    while (this.entries.size > this.config.capacityEntries) {
      const victim = [...this.entries.values()]
        .filter((entry) => entry.key !== protectedKey)
        .sort((left, right) => this.compareVictims(left, right))[0];
      if (!victim) break;
      this.entries.delete(victim.key);
      this.counters.evictions += 1;
      evicted.push(victim.key);
    }
    return evicted;
  }

  private compareVictims(left: Entry, right: Entry): number {
    if (this.config.evictionPolicy === EvictionPolicy.Lfu && left.accessCount !== right.accessCount) return left.accessCount - right.accessCount;
    if (left.accessSeq !== right.accessSeq) return left.accessSeq - right.accessSeq;
    if (left.accessedAt !== right.accessedAt) return left.accessedAt - right.accessedAt;
    if (left.createdAt !== right.createdAt) return left.createdAt - right.createdAt;
    return left.key.localeCompare(right.key);
  }

  private matches(scope: Invalidation, entry: Entry): boolean {
    if (scope.kind === 'key') return entry.key === scope.value;
    if (scope.kind === 'namespace') return entry.namespace === scope.value;
    return entry.key.startsWith(scope.value);
  }

  private nextAccessSeq(): number {
    this.accessSeq += 1;
    return this.accessSeq;
  }
}

export class HashRing {
  private tokens: Array<{ token: bigint; node: NodeInfo }> = [];
  version = 0;

  constructor(nodes: NodeInfo[], private readonly replicas: number) {
    for (const node of nodes) this.add(node);
  }

  add(node: NodeInfo): void {
    for (let index = 0; index < this.replicas; index += 1) {
      this.tokens.push({ token: hash(`${node.id}-${index}`), node });
    }
    this.tokens.sort((left, right) => (left.token < right.token ? -1 : left.token > right.token ? 1 : 0));
    this.version += 1;
  }

  owner(key: string): NodeInfo {
    const token = hash(key);
    return this.tokens.find((entry) => entry.token >= token)?.node ?? this.tokens[0]?.node ?? new NodeInfo('unavailable');
  }

  tokensFor(nodeId: string): bigint[] {
    return this.tokens.filter((entry) => entry.node.id === nodeId).map((entry) => entry.token);
  }
}

export class MemoryStore {
  private writesFail = false;

  constructor(private readonly data = new Map<string, string>()) {}

  load(key: string): string {
    const value = this.data.get(key);
    if (value === undefined) throw new CacheError('BACKING_STORE', 'backing store unavailable');
    return value;
  }

  write(key: string, value: string): void {
    if (this.writesFail) throw new CacheError('BACKING_STORE', 'backing store unavailable');
    this.data.set(key, value);
  }

  failWrites(value: boolean): void {
    this.writesFail = value;
  }
}

export type HttpResponse = { status: number; body: string };

export class HttpApp {
  constructor(private readonly cache: Cache) {}

  async handle(method: string, path: string, body = ''): Promise<HttpResponse> {
    if (path === '/health') return json(200, { status: 'ok' });
    if (path === '/metrics') return json(200, this.cache.metrics());
    if (path === '/cluster/ring') return json(200, this.cache.ringInfo());
    if (path === '/cache/invalidate' && method === 'POST') {
      const parsed = parseJson(body);
      const scopes = ['key', 'namespace', 'prefix'].filter((key) => typeof parsed[key] === 'string');
      if (scopes.length !== 1) return json(400, { code: 'INVALIDATION' });
      const scope = scopes[0] === 'key' ? Invalidation.key(String(parsed.key)) : scopes[0] === 'namespace' ? Invalidation.namespace(String(parsed.namespace)) : Invalidation.prefix(String(parsed.prefix));
      return json(202, { accepted: true, matchedApprox: await this.cache.invalidate(scope) });
    }
    if (!path.startsWith('/cache/')) return json(404, { code: 'NOT_FOUND' });
    const key = path.slice('/cache/'.length).split('?')[0] ?? '';
    try {
      if (method === 'GET') {
        const result = await this.cache.get(key, { loadOnMiss: path.includes('loadOnMiss=true') });
        return json(result.hit || result.loaded ? 200 : 404, result);
      }
      if (method === 'PUT') {
        const parsed = parseJson(body);
        const result = await this.cache.set(key, String(parsed.value ?? ''), { ttlMs: numberOrUndefined(parsed.ttlMs), namespace: stringOrUndefined(parsed.namespace), writeThrough: parsed.writeThrough === true });
        return json(result.version === 1 ? 201 : 200, result);
      }
      if (method === 'DELETE') return json(200, await this.cache.delete(key));
      return json(405, { code: 'METHOD_NOT_ALLOWED' });
    } catch (error) {
      if (error instanceof CacheError && error.code === 'VALUE_TOO_LARGE') return json(413, { code: error.code, message: error.message });
      return json(400, { code: error instanceof CacheError ? error.code : 'ERROR' });
    }
  }
}

function hash(value: string): bigint {
  const hex = createHash('sha256').update(value).digest('hex').slice(0, 16);
  return BigInt(`0x${hex}`);
}

function json(status: number, value: unknown): HttpResponse {
  return { status, body: JSON.stringify(value) };
}

function parseJson(body: string): Record<string, unknown> {
  const parsed: unknown = body.length > 0 ? JSON.parse(body) : {};
  return parsed && typeof parsed === 'object' && !Array.isArray(parsed) ? parsed as Record<string, unknown> : {};
}

function numberOrUndefined(value: unknown): number | undefined {
  return typeof value === 'number' ? value : undefined;
}

function stringOrUndefined(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}
